File: java/com/google/copybara/http/HttpModule.java
/*
 * Copyright (C) 2023 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.http;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableMap;
import com.google.copybara.CheckoutPath;
import com.google.copybara.EndpointProvider;
import com.google.copybara.Trigger;
import com.google.copybara.checks.Checker;
import com.google.copybara.config.SkylarkUtil;
import com.google.copybara.credentials.CredentialIssuer;
import com.google.copybara.credentials.CredentialModule.UsernamePasswordIssuer;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.http.auth.AuthInterceptor;
import com.google.copybara.http.auth.BearerInterceptor;
import com.google.copybara.http.auth.UsernamePasswordInterceptor;
import com.google.copybara.http.endpoint.HttpEndpoint;
import com.google.copybara.http.json.HttpEndpointJsonContent;
import com.google.copybara.http.multipart.FilePart;
import com.google.copybara.http.multipart.HttpEndpointFormPart;
import com.google.copybara.http.multipart.HttpEndpointMultipartFormContent;
import com.google.copybara.http.multipart.HttpEndpointUrlEncodedFormContent;
import com.google.copybara.http.multipart.TextPart;
import com.google.copybara.util.console.Console;
import java.net.URLEncoder;
import java.util.Map;
import java.util.Optional;
import javax.annotation.Nullable;
import net.starlark.java.annot.Param;
import net.starlark.java.annot.ParamType;
import net.starlark.java.annot.StarlarkBuiltin;
import net.starlark.java.annot.StarlarkMethod;
import net.starlark.java.eval.Dict;
import net.starlark.java.eval.EvalException;
import net.starlark.java.eval.NoneType;
import net.starlark.java.eval.Sequence;
import net.starlark.java.eval.StarlarkValue;

/** Starlark methods for working with the http endpoint. */
@StarlarkBuiltin(name = "http", doc = "Module for working with http endpoints.")
public class HttpModule implements StarlarkValue {
  HttpOptions options;
  Console console;

  public HttpModule(Console console, HttpOptions options) {
    this.console = console;
    this.options = options;
  }

  @StarlarkMethod(
      name = "url_encode",
      doc = "URL-encode the input string",
      parameters = {
        @Param(
            name = "input",
            doc = "The string to be encoded.",
            allowedTypes = {@ParamType(type = String.class)})
      })
  public String urlEncode(String input) {
    return URLEncoder.encode(input, UTF_8);
  }

  @StarlarkMethod(
      name = "trigger",
      doc = "Trigger for http endpoint",
      parameters = {
        @Param(
            name = "hosts",
            doc = "A list of hosts to allow HTTP traffic to.",
            named = true,
            allowedTypes = {@ParamType(type = Sequence.class)},
            defaultValue = "[]",
            positional = false),
        @Param(
            name = "issuers",
            doc = "A dictionary of credential issuers.",
            named = true,
            allowedTypes = {@ParamType(type = Dict.class), @ParamType(type = NoneType.class)},
            defaultValue = "{}",
            positional = false),
        @Param(
            name = "checker",
            allowedTypes = {
              @ParamType(type = Checker.class),
              @ParamType(type = NoneType.class),
            },
            defaultValue = "None",
            doc = "A checker that will check calls made by the endpoint",
            named = true,
            positional = false),
      })
  public Trigger trigger(Sequence<?> hosts, @Nullable Object issuers, @Nullable Object checkerIn)
      throws ValidationException, EvalException {
    HttpEndpoint endpoint =
        new HttpEndpoint(
            console,
            options.getTransport(),
            buildHostsMapWithAuthInterceptor(hosts).buildKeepingLast(),
            buildIssuersMap(issuers).buildKeepingLast(),
            SkylarkUtil.convertFromNoneable(checkerIn, null));
    return new HttpTrigger(endpoint);
  }

  @StarlarkMethod(
      name = "endpoint",
      doc =
          "Endpoint that executes any sort of http request. Currently restricted"
              + "to requests to specific hosts.",
      parameters = {
        @Param(
            name = "host",
            doc = "DEPRECATED. A single host to allow HTTP traffic to.",
            named = true,
            allowedTypes = {
              @ParamType(type = String.class),
            },
            defaultValue = "''",
            positional = false),
        @Param(
            name = "checker",
            allowedTypes = {
              @ParamType(type = Checker.class),
              @ParamType(type = NoneType.class),
            },
            defaultValue = "None",
            doc = "A checker that will check calls made by the endpoint",
            named = true,
            positional = false),
        @Param(
            name = "hosts",
            doc = "A list of hosts to allow HTTP traffic to.",
            named = true,
            allowedTypes = {@ParamType(type = Sequence.class)},
            defaultValue = "[]",
            positional = false),
        @Param(
            name = "issuers",
            doc = "A dictionaty of credential issuers.",
            named = true,
            allowedTypes = {@ParamType(type = Dict.class), @ParamType(type = NoneType.class)},
            defaultValue = "{}",
            positional = false),
      })
  public EndpointProvider<HttpEndpoint> endpoint(
      @Nullable String host,
      @Nullable Object checkerIn,
      Sequence<?> hosts,
      @Nullable Object issuers)
      throws ValidationException, EvalException {
    @Nullable Checker checker = SkylarkUtil.convertFromNoneable(checkerIn, null);
    ImmutableMap.Builder<String, Optional<AuthInterceptor>> h =
        buildHostsMapWithAuthInterceptor(hosts);
    if (host != null && !host.isEmpty()) {
      h.put(host, Optional.empty());
    }

    return EndpointProvider.wrap(
        new HttpEndpoint(
            console,
            options.getTransport(),
            h.buildKeepingLast(),
            buildIssuersMap(issuers).buildKeepingLast(),
            checker));
  }

  private ImmutableMap.Builder<String, CredentialIssuer> buildIssuersMap(Object issuers) {
    ImmutableMap.Builder<String, CredentialIssuer> issuersMap = ImmutableMap.builder();
    if (issuers == null) {
      return issuersMap;
    }
    for (Map.Entry<?, ?> entry : ((Dict<?, ?>) issuers).entrySet()) {
      issuersMap.put((String) entry.getKey(), (CredentialIssuer) entry.getValue());
    }
    return issuersMap;
  }

  private ImmutableMap.Builder<String, Optional<AuthInterceptor>> buildHostsMapWithAuthInterceptor(
      Sequence<?> hosts) {
    ImmutableMap.Builder<String, Optional<AuthInterceptor>> h = ImmutableMap.builder();
    for (Object o : hosts) {
      if (o instanceof HostCredential withCred) {
        h.put(withCred.host(), withCred.creds());
      } else {
        h.put((String) o, Optional.empty());
      }
    }
    return h;
  }

  @StarlarkMethod(
      name = "urlencoded_form",
      doc = "Creates a url-encoded form HTTP body.",
      parameters = {
        @Param(
            name = "body",
            doc = "HTTP body object, property name will be used as key and value as value.",
            allowedTypes = {@ParamType(type = Dict.class)},
            defaultValue = "{}"),
      })
  public HttpEndpointUrlEncodedFormContent urlEncodedFormContent(Object body) {
    return new HttpEndpointUrlEncodedFormContent(body);
  }

  @StarlarkMethod(
      name = "multipart_form",
      doc = "Creates a multipart form http body.",
      parameters = {
        @Param(
            name = "parts",
            doc = "A list of form parts",
            allowedTypes = {
              @ParamType(type = Sequence.class),
            },
            defaultValue = "[]")
      })
  public HttpEndpointMultipartFormContent multipartFormContent(Sequence<?> partsIn)
      throws EvalException {
    Sequence<HttpEndpointFormPart> parts =
        Sequence.cast(partsIn, HttpEndpointFormPart.class, "parts");
    return new HttpEndpointMultipartFormContent(parts);
  }

  @StarlarkMethod(
      name = "multipart_form_text",
      doc = "Create a text/plain part for a multipart form payload",
      parameters = {
        @Param(
            name = "name",
            doc = "The name of the form field.",
            allowedTypes = {@ParamType(type = String.class)}),
        @Param(
            name = "text",
            doc = "The form value of the field",
            allowedTypes = {@ParamType(type = String.class)})
      })
  public HttpEndpointFormPart multipartFormTextField(String name, String text) {
    return new TextPart(name, text);
  }

  @StarlarkMethod(
      name = "multipart_form_file",
      doc = "Create a file part for a multipart form payload.",
      parameters = {
        @Param(
            name = "name",
            doc = "The name of the form field.",
            allowedTypes = {
              @ParamType(type = String.class),
            }),
        @Param(
            name = "path",
            doc = "The checkout path pointing to the file to use " + "as the field value.",
            allowedTypes = {
              @ParamType(type = CheckoutPath.class),
            }),
        @Param(
            name = "content_type",
            doc =
                "Content type header value for the form part. "
                    + "Defaults to application/octet-stream. \n"
                    + "https://www.w3.org/Protocols/rfc1341/4_Content-Type.html",
            allowedTypes = {
              @ParamType(type = String.class),
            },
            named = true,
            positional = false,
            defaultValue = "\"application/octet-stream\""),
        @Param(
            name = "filename",
            doc =
                "The filename that will be sent along with the data. "
                    + "Defaults to the filename of the path parameter. "
                    + "Sets the filename parameter in the content disposition "
                    + "header. \n"
                    + "https://www.w3.org/Protocols/HTTP/Issues/content-disposition.txt",
            allowedTypes = {
              @ParamType(type = String.class),
              @ParamType(type = NoneType.class),
            },
            named = true,
            positional = false,
            defaultValue = "None"),
      })
  public HttpEndpointFormPart multipartFormFileField(
      String name, CheckoutPath path, String contentType, Object filenameIn) {
    @Nullable String filename = SkylarkUtil.convertOptionalString(filenameIn);
    return new FilePart(name, path.getCheckoutDir().resolve(path.getPath()), contentType, filename);
  }

  @StarlarkMethod(
      name = "json",
      doc = "Creates a JSON HTTP body.",
      parameters = {
        @Param(
            name = "body",
            doc = "HTTP body object, property name will be used as key and value as value.",
            allowedTypes = {@ParamType(type = Object.class)},
            defaultValue = "{}"),
      })
  public HttpEndpointJsonContent jsonContent(Object body) {
    return new HttpEndpointJsonContent(body);
  }

  @StarlarkMethod(
      name = "host",
      doc = "Wraps a host and potentially credentials for http auth.",
      parameters = {
        @Param(
            name = "host",
            doc = "The host to be contacted.",
            named = true,
            allowedTypes = {
              @ParamType(type = String.class),
            },
            positional = false),
        @Param(
            name = "auth",
            doc = "Optional, an interceptor for providing credentials. Also accepts a "
                + "username_password.",
            named = true,
            defaultValue = "None",
            allowedTypes = {
              @ParamType(type = AuthInterceptor.class),
              @ParamType(type = UsernamePasswordIssuer.class),
              @ParamType(type = NoneType.class)
            },
            positional = false)
      })
  public HostCredential host(String host, Object maybeCreds) {
    maybeCreds = maybeCreds instanceof UsernamePasswordIssuer
        ? new UsernamePasswordInterceptor((UsernamePasswordIssuer) maybeCreds) : maybeCreds;
    AuthInterceptor creds = SkylarkUtil.convertFromNoneable(maybeCreds, null);
    return new AutoValue_HttpModule_HostCredential(host, Optional.ofNullable(creds));
  }

  @StarlarkMethod(
      name = "username_password_auth",
      doc = "Authentication via username and password.",
      parameters = {
        @Param(
            name = "creds",
            doc = "The username and password credentials.",
            named = true,
            allowedTypes = {
              @ParamType(type = UsernamePasswordIssuer.class),
            },
            positional = false),
      })
  public UsernamePasswordInterceptor usernamePasswordAuth(UsernamePasswordIssuer creds) {
    return new UsernamePasswordInterceptor(creds);
  }

  @StarlarkMethod(
      name = "bearer_auth",
      doc = "Authentication via a bearer token.",
      parameters = {
        @Param(
            name = "creds",
            doc = "The token credentials.",
            named = true,
            allowedTypes = {
              @ParamType(type = CredentialIssuer.class),
            },
            positional = false),
      })
  public BearerInterceptor bearerAuth(CredentialIssuer creds) {
    return new BearerInterceptor(creds);
  }

  /** A username/password issuer pair tied to a host */
  @AutoValue
  public abstract static class HostCredential implements StarlarkValue {
    public abstract String host();

    public abstract Optional<AuthInterceptor> creds();
  }
}

File: src/Copybara.Core/Http/HttpModule.cs
/*
 * Copyright (C) 2023 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Text;
using Copybara.Checks;
using Copybara.Config;
using Copybara.Credentials;
using Copybara.Exceptions;
using Copybara.Http.Auth;
using Copybara.Http.Endpoint;
using Copybara.Http.Json;
using Copybara.Http.Multipart;
using Starlark.Annot;
using Starlark.Eval;
using Console = Copybara.Util.Console.Console;
using StarlarkList = Starlark.Eval.StarlarkList;

namespace Copybara.Http;

/// <summary>Starlark methods for working with the http endpoint.</summary>
[StarlarkBuiltin("http", Doc = "Module for working with http endpoints.")]
public class HttpModule : IStarlarkValue
{
    private readonly Console _console;
    private readonly HttpOptions _options;

    public HttpModule(Console console, HttpOptions options)
    {
        _console = console;
        _options = options;
    }

    [StarlarkMethod("url_encode", Doc = "URL-encode the input string")]
    public string UrlEncode(
        [Param(Name = "input", Doc = "The string to be encoded.",
            AllowedTypes = new[] { typeof(string) })]
        string input) => Uri.EscapeDataString(input);

    [StarlarkMethod("trigger", Doc = "Trigger for http endpoint")]
    public ITrigger Trigger(
        [Param(
            Name = "hosts",
            Doc = "A list of hosts to allow HTTP traffic to.",
            Named = true,
            AllowedTypes = new[] { typeof(StarlarkList) },
            DefaultValue = "[]",
            Positional = false)]
        StarlarkList hosts,
        [Param(
            Name = "issuers",
            Doc = "A dictionary of credential issuers.",
            Named = true,
            AllowedTypes = new[] { typeof(Dict), typeof(NoneType) },
            DefaultValue = "{}",
            Positional = false)]
        object? issuers,
        [Param(
            Name = "checker",
            AllowedTypes = new[] { typeof(IChecker), typeof(NoneType) },
            DefaultValue = "None",
            Doc = "A checker that will check calls made by the endpoint",
            Named = true,
            Positional = false)]
        object? checkerIn)
    {
        var endpoint = new HttpEndpoint(
            _console,
            _options.GetTransport(),
            BuildHostsMapWithAuthInterceptor(hosts),
            BuildIssuersMap(issuers),
            SkylarkUtil.ConvertFromNoneable<IChecker>(checkerIn, null));
        return new HttpTrigger(endpoint);
    }

    [StarlarkMethod(
        "endpoint",
        Doc = "Endpoint that executes any sort of http request. Currently restricted"
            + "to requests to specific hosts.")]
    public EndpointProvider<HttpEndpoint> Endpoint(
        [Param(
            Name = "host",
            Doc = "DEPRECATED. A single host to allow HTTP traffic to.",
            Named = true,
            AllowedTypes = new[] { typeof(string) },
            DefaultValue = "''",
            Positional = false)]
        string? host,
        [Param(
            Name = "checker",
            AllowedTypes = new[] { typeof(IChecker), typeof(NoneType) },
            DefaultValue = "None",
            Doc = "A checker that will check calls made by the endpoint",
            Named = true,
            Positional = false)]
        object? checkerIn,
        [Param(
            Name = "hosts",
            Doc = "A list of hosts to allow HTTP traffic to.",
            Named = true,
            AllowedTypes = new[] { typeof(StarlarkList) },
            DefaultValue = "[]",
            Positional = false)]
        StarlarkList hosts,
        [Param(
            Name = "issuers",
            Doc = "A dictionaty of credential issuers.",
            Named = true,
            AllowedTypes = new[] { typeof(Dict), typeof(NoneType) },
            DefaultValue = "{}",
            Positional = false)]
        object? issuers)
    {
        IChecker? checker = SkylarkUtil.ConvertFromNoneable<IChecker>(checkerIn, null);
        var h = BuildHostsMapBuilder(hosts);
        if (!string.IsNullOrEmpty(host))
        {
            h[host] = null;
        }

        return EndpointProvider.Wrap(
            new HttpEndpoint(
                _console,
                _options.GetTransport(),
                h.ToImmutableDictionary(),
                BuildIssuersMap(issuers),
                checker));
    }

    private ImmutableDictionary<string, CredentialIssuer> BuildIssuersMap(object? issuers)
    {
        var issuersMap = ImmutableDictionary.CreateBuilder<string, CredentialIssuer>();
        if (issuers is Dict dict)
        {
            foreach (var entry in dict.Entries)
            {
                issuersMap[(string)entry.Key!] = (CredentialIssuer)entry.Value!;
            }
        }

        return issuersMap.ToImmutable();
    }

    private ImmutableDictionary<string, IAuthInterceptor?> BuildHostsMapWithAuthInterceptor(
        StarlarkList hosts) => BuildHostsMapBuilder(hosts).ToImmutableDictionary();

    private Dictionary<string, IAuthInterceptor?> BuildHostsMapBuilder(StarlarkList hosts)
    {
        var h = new Dictionary<string, IAuthInterceptor?>();
        foreach (object? o in hosts)
        {
            if (o is HostCredential withCred)
            {
                h[withCred.Host] = withCred.Creds;
            }
            else
            {
                h[(string)o!] = null;
            }
        }

        return h;
    }

    [StarlarkMethod("urlencoded_form", Doc = "Creates a url-encoded form HTTP body.")]
    public HttpEndpointUrlEncodedFormContent UrlEncodedFormContent(
        [Param(
            Name = "body",
            Doc = "HTTP body object, property name will be used as key and value as value.",
            AllowedTypes = new[] { typeof(Dict) },
            DefaultValue = "{}")]
        object body) => new(body);

    [StarlarkMethod("multipart_form", Doc = "Creates a multipart form http body.")]
    public HttpEndpointMultipartFormContent MultipartFormContent(
        [Param(
            Name = "parts",
            Doc = "A list of form parts",
            AllowedTypes = new[] { typeof(StarlarkList) },
            DefaultValue = "[]")]
        StarlarkList partsIn)
    {
        var parts = Sequence.Cast<IHttpEndpointFormPart>(partsIn, "parts").ToList();
        return new HttpEndpointMultipartFormContent(parts);
    }

    [StarlarkMethod("multipart_form_text", Doc = "Create a text/plain part for a multipart form payload")]
    public IHttpEndpointFormPart MultipartFormTextField(
        [Param(Name = "name", Doc = "The name of the form field.",
            AllowedTypes = new[] { typeof(string) })]
        string name,
        [Param(Name = "text", Doc = "The form value of the field",
            AllowedTypes = new[] { typeof(string) })]
        string text) => new TextPart(name, text);

    [StarlarkMethod("multipart_form_file", Doc = "Create a file part for a multipart form payload.")]
    public IHttpEndpointFormPart MultipartFormFileField(
        [Param(Name = "name", Doc = "The name of the form field.",
            AllowedTypes = new[] { typeof(string) })]
        string name,
        [Param(Name = "path",
            Doc = "The checkout path pointing to the file to use as the field value.",
            AllowedTypes = new[] { typeof(CheckoutPath) })]
        CheckoutPath path,
        [Param(
            Name = "content_type",
            Doc = "Content type header value for the form part. "
                + "Defaults to application/octet-stream. \n"
                + "https://www.w3.org/Protocols/rfc1341/4_Content-Type.html",
            AllowedTypes = new[] { typeof(string) },
            Named = true,
            Positional = false,
            DefaultValue = "\"application/octet-stream\"")]
        string contentType,
        [Param(
            Name = "filename",
            Doc = "The filename that will be sent along with the data. "
                + "Defaults to the filename of the path parameter. "
                + "Sets the filename parameter in the content disposition "
                + "header. \n"
                + "https://www.w3.org/Protocols/HTTP/Issues/content-disposition.txt",
            AllowedTypes = new[] { typeof(string), typeof(NoneType) },
            Named = true,
            Positional = false,
            DefaultValue = "None")]
        object? filenameIn)
    {
        string? filename = SkylarkUtil.ConvertOptionalString(filenameIn);
        return new FilePart(name, path.FullPath(), contentType, filename);
    }

    [StarlarkMethod("json", Doc = "Creates a JSON HTTP body.")]
    public HttpEndpointJsonContent JsonContent(
        [Param(
            Name = "body",
            Doc = "HTTP body object, property name will be used as key and value as value.",
            AllowedTypes = new[] { typeof(object) },
            DefaultValue = "{}")]
        object? body) => new(body);

    [StarlarkMethod("host", Doc = "Wraps a host and potentially credentials for http auth.")]
    public HostCredential Host(
        [Param(
            Name = "host",
            Doc = "The host to be contacted.",
            Named = true,
            AllowedTypes = new[] { typeof(string) },
            Positional = false)]
        string host,
        [Param(
            Name = "auth",
            Doc = "Optional, an interceptor for providing credentials. Also accepts a "
                + "username_password.",
            Named = true,
            DefaultValue = "None",
            AllowedTypes = new[]
            {
                typeof(IAuthInterceptor),
                typeof(UsernamePasswordIssuer),
                typeof(NoneType),
            },
            Positional = false)]
        object? maybeCreds)
    {
        if (maybeCreds is UsernamePasswordIssuer upi)
        {
            maybeCreds = new UsernamePasswordInterceptor(upi);
        }

        IAuthInterceptor? creds = SkylarkUtil.ConvertFromNoneable<IAuthInterceptor>(maybeCreds, null);
        return new HostCredential(host, creds);
    }

    [StarlarkMethod("username_password_auth", Doc = "Authentication via username and password.")]
    public UsernamePasswordInterceptor UsernamePasswordAuth(
        [Param(
            Name = "creds",
            Doc = "The username and password credentials.",
            Named = true,
            AllowedTypes = new[] { typeof(UsernamePasswordIssuer) },
            Positional = false)]
        UsernamePasswordIssuer creds) => new(creds);

    [StarlarkMethod("bearer_auth", Doc = "Authentication via a bearer token.")]
    public BearerInterceptor BearerAuth(
        [Param(
            Name = "creds",
            Doc = "The token credentials.",
            Named = true,
            AllowedTypes = new[] { typeof(CredentialIssuer) },
            Positional = false)]
        CredentialIssuer creds) => new(creds);

    /// <summary>A host paired with optional auth credentials.</summary>
    public sealed class HostCredential : IStarlarkValue
    {
        public HostCredential(string host, IAuthInterceptor? creds)
        {
            Host = host;
            Creds = creds;
        }

        public string Host { get; }

        public IAuthInterceptor? Creds { get; }
    }
}
